fly/commands/internal/flaghelpers/pipeline_flag_test.go

```diff
@@ -17,52 +19,98 @@ var _ = Describe("PipelineFlag", func() {
-			})
-		})
-
-		Context("when an instance vars is complex", func() {
-			It("unmarshal the pipeline name and instance vars correctly", func() {
-				err := pipelineFlag.UnmarshalFlag("some-pipeline/foo.bar.baz:1,foo.bar.qux:2,bar.0:1,bar.1:\"2\"")
-
-				Expect(err).ToNot(HaveOccurred())
-				Expect(pipelineFlag.Name).To(Equal("some-pipeline"))
-				Expect(pipelineFlag.InstanceVars).To(Equal(atc.InstanceVars{
-					"bar": []interface{}{1, "2"},
-					"foo": map[string]interface{}{
-						"bar": map[string]interface{}{
-							"baz": 1,
-							"qux": 2,
+		for _, tt := range []struct {
+			desc         string
+			flag         string
+			name         string
+			instanceVars atc.InstanceVars
+			err          string
+		}{
+			{
+				desc: "name",
+				flag: "some-pipeline",
+				name: "some-pipeline",
+			},
+			{
+				desc:         "instance var",
+				flag:         "some-pipeline/branch:master",
+				name:         "some-pipeline",
+				instanceVars: atc.InstanceVars{"branch": "master"},
+			},
+			{
+				desc: "multiple instance vars",
+				flag: `some-pipeline/branch:master,list:[1, "2"],other:{foo:bar: 123}`,
+				name: "some-pipeline",
+				instanceVars: atc.InstanceVars{
+					"branch": "master",
+					"list":   []interface{}{json.Number("1"), "2"},
+					"other":  map[string]interface{}{"foo:bar": json.Number("123")},
+				},
+			},
+			{
+				desc: "quoted yaml brackets/braces",
+				flag: `some-pipeline/field1:{a: "{", b: "["},field2:hello`,
+				name: "some-pipeline",
+				instanceVars: atc.InstanceVars{
+					"field1": map[string]interface{}{
+						"a": "{",
+						"b": "[",
+					},
+					"field2": "hello",
+				},
+			},
+			{
+				desc: "whitespace trimmed from path/values",
+				flag: `some-pipeline/branch: master, other: 123`,
+				name: "some-pipeline",
+				instanceVars: atc.InstanceVars{
+					"branch": "master",
+					"other":  json.Number("123"),
+				},
+			},
+			{
+				desc: "quoted fields can contain special characters",
+				flag: `some-pipeline/"some.field:here":abc`,
+				name: "some-pipeline",
+				instanceVars: atc.InstanceVars{
+					"some.field:here": "abc",
+				},
+			},
+			{
+				desc: "supports dot notation",
+				flag: `some-pipeline/"my.field".subkey1."subkey:2":"my-value","my.field".other:"other-value"`,
+				name: "some-pipeline",
+				instanceVars: atc.InstanceVars{
+					"my.field": map[string]interface{}{
+						"subkey1": map[string]interface{}{
+							"subkey:2": "my-value",
 						},
+						"other": "other-value",
 					},
-				}))
+				},
+			},
+			{
+				desc: "errors if invalid ref is passed",
+				flag: `some-pipeline/"my.field".:bad`,
+				err:  `invalid var '"my.field".': empty field`,
+			},
+			{
+				desc: "errors if invalid YAML is passed as the value",
+				flag: `some-pipeline/hello:{bad: yaml`,
+				err:  `invalid value for key 'hello': error converting YAML to JSON: yaml: line 1: did not find expected ',' or '}'`,
+			},
+		} {
```

Comment:
Let's add some tests for special characters like `'pipeline/field:"value1,value2"'` and `'pipeline/field:"value1:value2"'`